_test/test_dnd_class/test_unary_ops.m

```diff
@@ -0,0 +1,93 @@
+classdef test_unary_ops < TestCase
+
+methods
+
+    function obj = test_unary_ops(~)
+        obj = obj@TestCase('test_unary_ops');
+    end
+
+    function test_all_functions_are_defined(obj)
+        % the unary operation and the range the data it acts on should take
+        unary_ops = {
+            @acos, [0, 1], ...
+            @acosh, [1, 3], ...
+            @acot, [0, 1], ...
+            @acoth, [10, 15], ...
+            @acsc, [1, 3], ...
+            @acsch, [1, 3], ...
+            @asec, [1.5, 3], ...
+            @asech, [0, 1], ...
+            @asin, [0, 1], ...
+            @asinh, [1, 3], ...
+            @atan, [0, 1], ...
+            @atanh, [0, 0.5], ...
+            @cos, [0, 1], ...
+            @cosh, [0, 1], ...
+            @cot, [0, 1], ...
+            @coth, [1.5, 3], ...
+            @csc, [0.5, 2.5], ...
+            @csch, [1, 3], ...
+            @exp, [0, 1], ...
+            @log, [1, 3], ...
+            @log10, [1, 3], ...
+            @sec, [2, 4], ...
+            @sech, [0, 1.4], ...
+            @sin, [0, 3], ...
+            @sinh, [0, 3], ...
+            @sqrt, [0, 3], ...
+            @tan, [0, 1], ...
+            @tanh, [0, 3], ...
+        };
+
+        % For each unary operator, perform the operation on some data
+        num_pix = 7;
+        for i = 1:2:numel(unary_ops)
+            unary_op = unary_ops{i};
+            data_range = unary_ops{i+1};
+
+            data.s = obj.get_random_data_in_range( ...
+                PixelData.DEFAULT_NUM_PIX_FIELDS, num_pix, data_range);
+            data.e = obj.get_random_data_in_range( ...
+                PixelData.DEFAULT_NUM_PIX_FIELDS, num_pix, data_range);
+
+            dnd_obj = d2d();
+            dnd_obj.npix = [num_pix];
+            dnd_obj.s = data.s;
+            dnd_obj.e = data.e;
+
+            % exception will be thrown if method not implemented
+            result = unary_op(dnd_obj);
+
+            % confirm the data (s, e) has been updated
+            assertFalse(equal_to_tol(result.s, dnd_obj.s));
```

Comment:
Would it be a better test of whether the functions are defined to operate with the identity and check that things are unchanged? 

If I understand correctly, currently you're testing random data which may happen to be the identity, which would result in false positives. 

Personally, I don't like random data in tests in general.

Reply:
I agree that the use of random data in tests is something to consider carefully.


The prime purpose of this test is that the function exists in the class API and does not generate an error when called with an appropriately shaped data array.

In calling it with an `nxm` array of random numbers that lie within "valid input range" for the operations (so we're not, for example, asking for logs of negative numbers) the likelihood of all of these values being the identity value of the function is vanishingly small.

The second test verifies the actual implementation of "an example" unary function to confirm the resulting numerical data is correct.

Reply:
This is a cut-down implementation of the sqw unary operator tests where there is an additional layer of complexity in the function is applied to the raw pixel data (data.pix.s) and the image data is recalculated from that (data.s). There are similar tests in the PixelData class too.
Some simplification is needed in these tests and will be made in a new commit.

Comprehensive testing of the actual mathematical implementations of the various unary ops should be done in Herbert where the unary operators are implemented. Those tests are currently missing.